lib/dockerregistry/checks.go

```diff
@@ -0,0 +1,169 @@
+	masterSHAStr := os.Getenv("PULL_BASE_SHA")
+	err := isValidSha(masterSHAStr)
+	if err != nil {
+		return nil, fmt.Errorf("The PULL_BASE_SHA environment variable "+
+			"is invalid: %v", err)
+	}
+	pullRequestSHAStr := os.Getenv("PULL_PULL_SHA")
+	err = isValidSha(pullRequestSHAStr)
+	if err != nil {
+		return nil, fmt.Errorf("The PULL_PULL_SHA environment variable "+
+			"is invalid: %v", err)
+	}
+	masterSHA := plumbing.NewHash(masterSHAStr)
+	pullRequestSHA := plumbing.NewHash(pullRequestSHAStr)
+	return &ImageRemovalCheck{
+		gitRepoPath,
+		masterSHA,
+		pullRequestSHA,
+	}, nil
+}
+
+// Run executes ImageRemovalCheck on a set of promotion edges.
+// Returns an error if the pull request removes images from the
+// promoter manifests.
+func (check *ImageRemovalCheck) Run(
+	edges map[PromotionEdge]interface{},
+) error {
+	var err error
+	// Open the Git repo of the current working directory, check
+	// out the master branch of the repo, parse and generate promotion
+	// edges from the manifests, and then reset the working directory
+	// by checking out the pull request branch of the Git repo. This
+	// is done to make sure that this check is not leaving any lasting
+	// effects on the working directory when the check finishes.
+	r, err := gogit.PlainOpen(check.GitRepoPath)
+	if err != nil {
+		return fmt.Errorf("Could not open the Git repo: %v", err)
+	}
+
+	// Check out the master branch of the repo
+	w, err := r.Worktree()
+	if err != nil {
+		return fmt.Errorf("Could not create Git worktree: %v", err)
+	}
+	err = w.Checkout(&gogit.CheckoutOptions{
+		Hash:  check.MasterSHA,
+		Force: true,
+	})
+	if err != nil {
+		return fmt.Errorf("Could not checkout the master branch of the Git"+
+			" repo: %v", err)
+	}
+
+	// Parse the manifests and generate the promotion edges
+	mfests, err := ParseThinManifestsFromDir(check.GitRepoPath)
+	if err != nil {
+		return fmt.Errorf("Could not parse manifests from the directory: %v",
+			err)
+	}
+	masterEdges, err := ToPromotionEdges(mfests)
+	if err != nil {
+		return fmt.Errorf("Could not generate promotion edges from promoter"+
+			" manifests: %v", err)
+	}
+
+	// Check out the pull request branch of the repo
+	err = w.Checkout(&gogit.CheckoutOptions{
+		Hash:  check.PullRequestSHA,
+		Force: true,
+	})
+	if err != nil {
+		return fmt.Errorf("Could not checkout the pull request branch of the"+
+			" Git repo %v: %v",
+			check.GitRepoPath, err)
+	}
+
+	return check.Compare(masterEdges, edges)
+}
+
+// Compare is a function of the ImageRemovalCheck that handles
+// the comparison of the pull requests's set of promotion edges and
+// the master branch's set of promotion edges.
+func (check *ImageRemovalCheck) Compare(
+	edgesMaster map[PromotionEdge]interface{},
+	edgesPullRequest map[PromotionEdge]interface{},
+) error {
+	// Generate a set of all destination images that appear in
+	// the pull request's set of promotion edges
+	destinationImages := make(map[PromotionEdge]interface{})
```

Comment:
This is better, but I wonder if it's cleaner to just clear the unused parts in the edgesMaster and edgesPullRequest directly... any pros/cons you can think of?